### PR TITLE
bpo-32972: Async test case

### PR DESCRIPTION
**Pull request by asvetlov**
Add explicit `asyncSetUp` and `asyncTearDown` methods.
The rest is the same as for #13228

`AsyncTestCase` create a loop instance for every test for the sake of test isolation.
Sometimes a loop shared between all tests can speed up tests execution time a lot but it requires control of closed resources after every test finish. Basically, it requires nested supervisors support that was discussed with @1st1 many times. Sorry, asyncio supervisors have no chance to land on Python 3.8.

The PR intentionally does not provide API for changing the used event loop or getting the test loop: use `asyncio.set_event_loop_policy()` and `asyncio.get_event_loop()` instead.

The PR adds four overridable methods to base `unittest.TestCase` class:
```
    def _callSetUp(self):
        self.setUp()

    def _callTestMethod(self, method):
        method()

    def _callTearDown(self):
        self.tearDown()

    def _callCleanup(self, function, /, *args, **kwargs):
        function(*args, **kwargs)
```
It allows using asyncio facilities with minimal influence on the unittest code.

The last but not least: the PR respects contextvars. The context variable installed by `asyncSetUp` is available on test, `tearDown` and a coroutine scheduled by `addCleanup`.

https://bugs.python.org/issue32972

### CODE REVIEW

File: Lib/unittest/async_case.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+    def _callTestMethod(self, method):
+        self._callMaybeAsync(method)
+
+    def _callTearDown(self):
+        self._callAsync(self.asyncTearDown)
+        self.tearDown()
+
+    def _callCleanup(self, function, *args, **kwargs):
+        self._callMaybeAsync(function, *args, **kwargs)
```
Comment (1st1) [inline Lib/unittest/async_case.py:72]: Why don't we have an explicit `addAsyncCleanup`?

Comment (1st1) [reply Lib/unittest/async_case.py:72]: For example, we have an explicit async method in AsyncExitStack: https://docs.python.org/3/library/contextlib.html#contextlib.AsyncExitStack.push_async_callback

Even though we toyed with the idea of just reusing the push_callback method and checking if the returned value is an awaitable.

I'm more in favour of `addAsyncCleanup` in this case.

File: Lib/unittest/async_case.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
+import asyncio
+import inspect
+
+from .case import TestCase
+
+
+
+class AsyncioTestCase(TestCase):
```
Comment (1st1) [inline Lib/unittest/async_case.py]: Can we call it `IsolatedAsyncioTestCase` to signify that the loop is not shared between the tests?

File: Lib/unittest/async_case.py (hunk 1 above)
Comment (asvetlov) [reply Lib/unittest/async_case.py:72]: I can add `addAsyncCleanup()` method, it makes sense.
The implementation should push a callback into the same cleanup queue as used for `addCleanup()` though, to keep cleanups order.
I think running all sync cleanups before asyncs and vice versa is a bad idea.
Cleanups should be executed exactly in the reverse adding order.
Also, `self.doCleanups()` should execute all sync and async cleanups altogether.
Do you agree?

File: Lib/unittest/async_case.py (hunk 2 above)
Comment (asvetlov) [reply Lib/unittest/async_case.py]: We can.
I have a hope that non-isolated tests can reuse the same test case but require a little different test runner maybe (to keep a loop during the whole test run session).

Anyway, I'm ok with renaming the class right now and see what's happen when we add a shared loop environment.

Comment (asvetlov) [reply Lib/unittest/async_case.py]: Renamed.
Not sure if I like it too much (the long name becomes even longer) but can live with it, sure.

**Comment by auvipy** [comment]
this would be great to see in 3.8

File: Lib/unittest/async_case.py (hunk 1 above)
Comment (1st1) [reply Lib/unittest/async_case.py:72]: Yes.

**Review by 1st1 — APPROVED** [review]
Please add `addAsyncCleanup()` before merging.

File: Lib/unittest/async_case.py (hunk 2 above)
Comment (1st1) [reply Lib/unittest/async_case.py]: I don't like it too much either,  but the predictability of the behaviour of this new base class comes first.

**Comment by asvetlov** [comment]
`addAsyncCleanup()` is added

**Review by auvipy — APPROVED** [review]

**Comment by abetkin** [comment]
Hi all, I tested unittest in 3.8 and what I see is tests are run not concurrently: I have 5 tests that sleep for 3 seconds and the total time of execution is 15 seconds. What is the reason behind this?

**Pull request merged**